account_invoice_margin_sale/hooks.py

```diff
@@ -0,0 +1,51 @@
+def post_init_hook(cr, registry):
+    # update purchase_price for invoices from sale_order
+    cr.execute(
+        """
+        UPDATE account_move_line AS aml
+        SET purchase_price = (
+            SELECT COALESCE(SUM(order_line.purchase_price), 0.0)
+            FROM sale_order_line AS order_line
+            INNER JOIN sale_order AS so ON order_line.order_id = so.id
+            INNER JOIN sale_order_line_invoice_rel AS soi
+            ON soi.order_line_id = order_line.id
+            WHERE soi.invoice_line_id = aml.id
+        )
+        WHERE EXISTS (
+            SELECT 1
+            FROM sale_order_line AS order_line
+            INNER JOIN sale_order_line_invoice_rel AS soi
+            ON soi.order_line_id = order_line.id
+            WHERE soi.invoice_line_id = aml.id
+        );
```

Comment:
Please don't complicate SQL:

```suggestion
        UPDATE account_move_line AS aml
        SET purchase_price = sol.purchase_price,
            margin = ...
        FROM sale_order_line sol
        INNER JOIN sale_order_line_invoice_rel AS rel
        ON rel.order_line_id = sol.id
        WHERE rel.invoice_line_id = aml.id
```

and do something similar for the other one.